src/main/daml/Daml/Finance/Instrument/Swap/Fpml/Util.daml

```diff
@@ -615,30 +617,31 @@ calculateFloatingPaymentClaimsFromSwapStream floatingRateCalculation s
                 if pp.rateFixingDate < pp.calculationPeriod.adjustedEndDate
                 then error "fixing of notional must not occur before the final calculation date"
                 else
-                  when (TimeGte pp.rateFixingDate) $ scale notional $
-                  scale (convertImplicitDcfToActualDcf) $
-                  when (TimeGte p.adjustedEndDate) $ one currency
+                  scale (pp.notional * convertImplicitDcfToActualDcf) $ one currency
+              rateStart = ObserveAt floatingRateCalculation.floatingRateIndex
+                pp.calculationPeriod.adjustedStartDate
+              rateEnd = ObserveAt floatingRateCalculation.floatingRateIndex
+                pp.calculationPeriod.adjustedEndDate
               claim =
                 -- calculate a daily compounded reference rate using the corresponding index,
                 -- see page 36 here:
                 -- https://www.newyorkfed.org/medialibrary/Microsites/arrc/files/2021/users-guide-to-sofr2021-update.pdf
-                scaleByAssetPerformance pp.calculationPeriod.adjustedStartDate
-                  pp.calculationPeriod.adjustedEndDate rate endClaim
+                when (TimeGte p.adjustedEndDate) $ scale (rateEnd / rateStart - Const 1.0) endClaim
 
       -- Calculate a floating rate correction claim for a payment period
       -- (corresponding to the "-1" in a perf calculation)
       -- This will no longer be needed when improved compounding logic is supported in CC
       createCorrectionClaim p =
```

Comment:
In what case do we still need the correction claim?

Reply:
It comes from rate compounding over the calculation periods in one payment period, on the form r = (1+r1)(1+r2)(1+r3) - 1:

`          compoundRates acc pp =
            scale (Const 1.0 + rate * pp.dcf) acc
            where
              regularRateExclSpread = ObserveAt floatingRateCalculation.floatingRateIndex
                pp.rateFixingDate
              regularRateInclSpread = regularRateExclSpread + Const floatingRateSpread
              regularRate = if includeSpread then regularRateInclSpread else regularRateExclSpread
              rate = getFloatingRate s pp.calculationPeriod p regularRate rateFixingCalendar fixLeg
                pp.rateFixingDate

          compoundedRates = when (TimeGte p.adjustedEndDate) $ foldl
            compoundRates
            (one currency)
            (reverse paymentPeriodData)
`

where the correction claim corresponds to the -1.

I wonder whether it is possible to get rid of that? If so, do I need to put the whole compounding calculation in one `scale` node, i.e. create only the compounding expression (without `scale`) using `fold` and then put the `scale` around the result (including -1)?

Reply:
I have implemented it in this way now.